pyfr/plugins/integrate.py

```diff
@@ -0,0 +1,156 @@
+
+                # Get the smats at the solution points
+                smat = eles.smat_at_np('upts').transpose(2, 0, 1, 3)
+
+                # Get |J|^-1 at the solution points
+                rcpdjac = eles.rcpdjac_at_np('upts')
+
+                # Product to give J^-T at the solution points
+                self._rcpjact.append(smat*rcpdjac)
+
+    def _eval_exprs(self, intg):
+
+        # Get the primitive variable names
+        pnames = self.elementscls.privarmap[self.ndims]
+
+        exprs = np.zeros(len(self.exprs))
+        # Iterate over each element type in the simulation
+        for i, (soln, (plocs, rjacs, wts)) in enumerate(zip(intg.soln, self.elminfo)):
+    
+            # Convert from conservative to primitive variables
+            psolns = self.elementscls.con_to_pri(soln.swapaxes(0, 1),
+                                                 self.cfg)
+
+            # Prepare the substitutions dictionary
+            subs = dict(zip(pnames, psolns))
+            subs.update(zip('xyz', plocs.swapaxes(0, 1)))
+
+            # Compute any required gradients
+            if self._gradpnames:
+                # Gradient operator and J^-T matrix
+                gradop, rcpjact = self._gradop[i], self._rcpjact[i]
+                nupts = gradop.shape[1]
+
+                for pname in self._gradpnames:
+                    psoln = subs[pname]
+
+                    # Compute the transformed gradient
+                    tgradpn = gradop @ psoln
+                    tgradpn = tgradpn.reshape(self.ndims, nupts, -1)
+
+                    # Untransform this to get the physical gradient
+                    gradpn = np.einsum('ijkl,jkl->ikl', rcpjact, tgradpn)
+                    gradpn = gradpn.reshape(self.ndims, nupts, -1)
+
+                    for dim, grad in zip('xyz', gradpn):
+                        subs['grad_{0}_{1}'.format(pname, dim)] = grad
+
+            for j,(k,v) in enumerate(self.exprs):
+                # Accumulate integrated evaluated expressions
+                exprs[j] += np.sum(wts[:, None]*rjacs*npeval(v, subs))
+            
+            
+        # Stack up the expressions for each element type and return
+        return exprs
+
+    def __call__(self, intg):
+        doiint = intg.nacptsteps % self.nsteps == 0
+
+        if doiint:
+            # MPI info
+            comm, rank, root = get_comm_rank_root()
+            
+            # Evaluate the integation expressions
+            iintex = self._eval_exprs(intg)
+
+            # Reduce and output if we're the root rank
+            if rank != root:
+                comm.Reduce(iintex, None, op=get_mpi('sum'), root=root)
+            else:
+                comm.Reduce(get_mpi('in_place'), iintex, op=get_mpi('sum'), root=root)
+
+                # Build the row
+                row = [intg.tcurr] + iintex.tolist()
+
+                # Write
+                print(','.join(str(r) for r in row), file=self.outf)
```

Comment:
Maybe `print(intg.tcurr, *iintex.tolist(), sep=',', file=self.outf)` will do the job here?